File: P2-ListADT-c/List.c
//-----------------------------------------------------------------------------
// List.c
// Kevin Duong
// keduong
// 1-27-17
// Implementation file for List ADT
// Based off Stack.c from examples
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data, Node prev, Node next){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = next;
   N->prev = prev;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}


// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) { 
      Node N = (*pL)->front;
      while(N != NULL){
         Node clr = N;
         N = N->next;
         free(clr);
      }
      free(*pL);
      *pL = NULL;
   }
}

// Access functions -----------------------------------------------------------

// length()
// returns the number of elements in the List
int length(List L){
   if(L == NULL){
      printf("List Error: called length() on empty List");
      exit(1);
   }
   return L->length;
}

// index()
// Returns the index of the cursor, or -1
int index(List L){
   if(L == NULL){
      printf("List Error: called index() on NULL reference");
      exit(1);
   }
   return L->index;
}

// front()
// Returns the value at the front of L.
// Pre: length()>0
int front(List L){
   if( L==NULL ){
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling front() on an empty List\n");
      exit(1);
   }
   return(L->front->data);
}

// back()
// Returns back element
// Pre: length()>0
int back(List L){
   if( L==NULL ){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling back() on an empty List\n");
      exit(1);
   }
   return(L->back->data);
}
// get()
// Returns cursor element
// Pre: length()>0, index()>=0
int get(List L){
   if( L==NULL ){
      printf("List Error: calling get() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling get() on an empty List\n");
      exit(1);
   }
   if(L->index<0){
      printf("List Error: calling get() on undef reference\n");
      exit(1);
   }
   return(L->cursor->data);
}

// equals()
// Checks if List A and List B are equivalent
int equals(List A, List B){
   if(A == NULL || B == NULL) {
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   if(A->length != B->length) {
      return 0;
   }
   Node C = A->front;
   Node D = B->front;
   while(C->next != NULL && D->next != NULL) {
      if(C->data != D->data)
         return 0;
      C = C->next;
      D = D->next;
   }
   return 1;
}

// Manipulation procedures ----------------------------------------------------

// clear()
// Resets list to empty state
void clear(List L){
   if (L ==NULL){
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   Node N = L->front;
   while(N != NULL){
      Node temp = N;
      free(temp);
      N = N->next;
   }
   L->front = NULL;
   L->back = NULL;
   L->cursor = NULL;
   L->length = 0;
}

// moveFront()
// places cursor under the front element, or nothing
void moveFront(List L){
   if (L ==NULL){
      printf("List Error: calling moveFront() on NULL List reference\n");
      exit(1);
   }
   if(L->length >=1){
      L->cursor = L->front;
      L->index = 0;
   }
}

// moveBack()
// Places cursor under back element, or nothing
void moveBack(List L){
   if (L ==NULL){
      printf("List Error: calling moveBack() on NULL List reference\n");
      exit(1);
   }
   //int N = L->length - 1;
   if(L->length >=1){
      L->cursor = L->back;
      L->index = L->length - 1;
   }
}
// movePrev()
// Move cursor one step forward towards the front
void movePrev(List L) {
   if(L == NULL) {
      printf("List Error: calling movePrev() on NULL List reference\n");
      exit(1);
   }
   if(L->index != 0) {
      L->cursor = L->cursor->prev;
      --L->index;
   }
   else{
      L->cursor = NULL;
      L->index = -1;
   }
}

// moveNext()
// Move cursor one step forward towards the back
void moveNext(List L) {
   if(L == NULL) {
      printf("List Error: calling moveNext() on NULL List reference\n");
      exit(1);
   }
   int N = L->length -1;
   if(L->index != N) {
      L->cursor = L->cursor->next;
      ++L->index;
   }
   else{
      L->cursor = NULL;
      L->index = -1;
   }
}

// prepend()
// Insert new element into the list after the front
void prepend(List L, int data){
   if (L ==NULL){
      printf("List Error: calling prepend() on NULL List reference\n");
      exit(1);
   }
   Node N = newNode(data, NULL, L->front);
   if(L->front == NULL)
      L->back = L->front = N;
   else
      L->front = L->front->prev = N;
   L->length++;
}

// append()
// Insert new element into the list after the back
void append(List L, int data){
   if (L ==NULL){
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   Node N = newNode(data, L->back, NULL);
   if(L->front == NULL)
      L-> back = L->front = N;
   else
      L-> back = L->back->next = N;
   L->length++;
}

// insertBefore()
// Insert new element before cursor
// pre: length()>0; index()>=0
void insertBefore(List L, int data){
   if( L==NULL ){
      printf("List Error: calling insertBefore() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling insertBefore() on an empty List\n");
      exit(1);
   }
   if(L->index<0){
      printf("List Error: calling insertBefore() on undef reference\n");
      exit(1);
   }
   Node N = newNode(data, L->cursor->prev, L->cursor);
   if(L->cursor->prev != NULL)
      L->cursor->prev = L->cursor->prev->next= N;
   else
      L->cursor->prev = L->front = N;
   L->length++;
}

// insertAfter()
// Insert new element after cursor
// pre: length()>0; index()>=0
void insertAfter(List L, int data){
   if( L==NULL ){
      printf("List Error: calling insertAfter() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling insertAfter() on an empty List\n");
      exit(1);
   }
   if(L->index<0){
      printf("List Error: calling insertAfter() on undef reference\n");
      exit(1);
   }
   Node N = newNode(data, L->cursor, L->cursor->next);
   if(L->cursor->next != NULL)
      L->cursor->next = L->cursor->next->prev= N;
   else
      L->cursor->next = L->back = N;
   L->length++;
}

// deleteFront()
// Delete the front element.
// Pre: length()>=0
void deleteFront(List L){
   if( L==NULL ){
      printf("List Error: calling deleteFront() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling deleteFront() on an empty List\n");
      exit(1);
   }
   Node N;
   if(L->length > 1){ 
      N = L->front;
      L->front = L->front->next;
      L->front->prev = NULL;
   }
   else{
      N = L->front;
      L->front = L->back = NULL;
      L->cursor = NULL;
   }
   freeNode(&N);
   L->length++;
}
   
// deleteBack()
// Delete the back element.
// Pre: length()>=0
void deleteBack(List L){
   if( L==NULL ){
      printf("List Error: calling deleteBack() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling deleteBack() on an empty List\n");
      exit(1);
   }
   Node N;
   if(L->length > 1){ 
      N = L->back;
      L->back = L->back->prev;
      L->back->next = NULL;
   }
   else{
      N = L->back;
      L->back = L->front = NULL;
      L->cursor = NULL;
   }
   freeNode(&N);
   L->length++;
}
// delete()
// Delete cursor element
// Pre: length()>0, index()>=0
void delete(List L){
   if( L==NULL ){
      printf("List Error: calling delete() on NULL List reference\n");
      exit(1);
   }
   if( L->length <= 0 ){
      printf("List Error: calling delete() on an empty List\n");
      exit(1);
   }
   if(L->index<0){
      printf("List Error: calling delete() on undef reference\n");
      exit(1);
   }
   Node N = L->cursor->prev;
   N->next = L->cursor->next;
   if(N != NULL){
      L->cursor = L->cursor->next;
      L->cursor->prev = N;
   }
   L->length--;
}


// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in S on a single line to stdout.
void printList(FILE* out, List L){
   if( L==NULL ){
      fprintf(out,"Queue Error: calling printList() on NULL List reference\n");
      exit(1);
   }

   Node N = L->front;
   while(N != NULL){
      fprintf(out, "%d ", N->data);
      N = N->next;
   }
}

// copyList()
// Returns new List with same int. sequence as this list.
List copyList(List L){
   List copy = newList();
   Node N = L->front;
   while(N != NULL){
      append(copy, N->data);
      N = N->next;
   }
   return copy;
}

File: P1-ListADT-c/List.java
// List.java
// Kevin Duong 
// keduong
// 1-17-17
// PA1 - List.java
// List ADT for PA1
// Used Queue.java and Stack.java as framework

class List{

   private class Node{
      // Fields
      int data;
      Node next;
      Node prev;
      
      // Constructor
      Node(int data) { 
         this.data = data; 
         next = null; 
         prev = null; 
      }
      
      Node(int data, Node prev, Node next){
         this.data = data; 
         this.next = next; 
         this.prev = prev; 
      }
      
      // toString():  overrides Object's toString() method
      public String toString() { 
         return String.valueOf(data); 
      }
      
      // equals(): overrides Object's equals() method
      public boolean equals(Object x){
         boolean eq = false;
         Node that;
         if(x instanceof Node){
            that = (Node) x;
            eq = (this.data==that.data);
         }
         return eq;
      }
   }

   // Fields
   private Node front;
   private Node back;
   private Node cursor;
   private int index;
   private int length;

   // Constructor
   List() { 
      front = back = cursor = null; 
      length = 0; 
      index = -1;
   }


   // Access Functions --------------------------------------------------------

   // length()
   // Returns number of elements in this List
   int length() { 
      return length; 
   }
   
   // index()
   //  If cursor is defined, returns the index of the cursor element,
   //   otherwise returns -1.
   int index() { 
      return index; 
   }

   // front() 
   // Returns front element. 
   // Pre: length()>0
   int front(){
      if( length < 1 ){
         throw new RuntimeException(
            "List Error: front() called on empty List");
      }
      return front.data;
   }
   
   // back() 
   // Returns back element. 
   // Pre: length()>0
   int back(){
      if( length < 1 ){
         throw new RuntimeException(
            "List Error: back() called on empty List");
      }
      return back.data;
   }
   
   // get() 
   // Returns front element. 
   // Pre: length()>0, index()>= 0 
   int get(){
      if( length < 1 || index<0){
         throw new RuntimeException(
            "List Error: front() called on emptyList or undef. index");
      }
      return cursor.data;
   }
   
   // equals(List L)
   // Returns true if this List and L are the same integer
   // sequence. The cursor is ignored in both lists.
   boolean equals(List L){
      boolean eq = false;
      Node N, M;
      
      N = front;
      M = L.front;
      eq = (length == L.length);
      if (!eq) return eq;
      while(N != null && M != null){
         eq = N.equals(M);
         if (!eq) return eq;
         N = N.next;
         M = M.next;
      }
      return eq;
   }


   // Manipulation Procedures -------------------------------------------------

   // clear()
   // Resets this List to original empty State
   void clear(){
      if (front == null) return; //already empty
      front = front.next = null;
      length = 0;
   }

   // moveFront()
   // If List is non-empty, places the cursor under the front element,
   // otherwise does nothing.
   void moveFront(){
      if (length > 0){
         cursor = front;
         index = 0;
      }
   }
  
   // moveBack()
   // If List is non-empty, places the cursor under the back element,
   // otherwise does nothing.
   void moveBack(){
      if (length > 0){
         cursor = back;
         index = length-1;
      }
   }
   
   // movePrev()
   // If cursor is defined and not at front, moves cursor one step toward
   // front of this List, if cursor is defined and at front, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void movePrev(){
      if (cursor != null && index != 0){
         cursor = cursor.prev;
         --index;
      }
      else if(cursor != null && index == 0){
         cursor = null;
         index = -1;
      }
   }

   // moveNext()
   // If cursor is defined and not at back, moves cursor one step toward
   // back of this List, if cursor is defined and at back, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void moveNext(){
      if (cursor != null && index != length-1){
         cursor = cursor.next;
         ++index;
      }
      else if(cursor != null && index == length-1){
         cursor = null;
         index = -1;
      }
   }
   
   // prepend(int data)
   // Insert new element into this List. If List is non-empty,
   // insertion takes place before front element.
   void prepend(int data){
      Node N = new Node(data, null, front);
      ++length;
      if(front == null)
         back = front = N;
      else{
         front = front.prev = N;
      }
   }
   
   // append(int data)
   // Insert new element into this List. If List is non-empty,
   // insertion takes place before back element.
   void append(int data){
      Node N = new Node(data, back, null);
      ++length;
      if(front == null)
         front = back = N;
      else{
         back = back.next = N;
      }
   }
   
   //insertBefore(int Data)
   // Insert new element before cursor.
   // Pre: length()>0, index()>=0
   void insertBefore(int data){
      if (index <0 || length <1)
         throw new RuntimeException(
         "List Error: insertBefore() called on undefined index or empty List");
      Node N = new Node(data, cursor.prev, cursor);
      ++length;
      if (cursor.prev == null)
         front = cursor.prev = N;
      else{
         cursor.prev.next = N;
         cursor.prev = N;
      }
   }
   
   //insertAfter(int Data)
   // Insert new element after cursor.
   // Pre: length()>0, index()>=0
   void insertAfter(int data){
      if (index <0 || length <1)
         throw new RuntimeException(
         "List Error: insertBefore() called on undefined index or empty List");
      Node N = new Node(data, cursor, cursor.next);
      ++length;
      if (cursor.next == null)
         back = cursor.next = N;
      else{
         cursor.next.prev = N;
         cursor.next = N;
      }
   }
   
   // deleteFront()
   // Deletes the front element. 
   // Pre: length()>0   
   void deleteFront(){
      if (length<1)
         throw new RuntimeException(
      "List Error: deleteFront() called on empty list");
      if (cursor == front){
         front = back = null;
      }
      else{
         front = front.next;
         front.prev = null;
      }
      --length;
  
   }

   // deleteBack()
   // Deletes the front element. 
   // Pre: length()>0
   void deleteBack(){
      if (length<1)
         throw new RuntimeException(
      "List Error: deleteFront() called on empty list");
      back = back.prev;
      back.next = null;
   }
   
   // delete()
   // Deletes the cursor element, making the cursor undef.
   // Pre: length() >0, index() >=0
   void delete(){
      if (index <0 || length <1)
         throw new RuntimeException(
         "List Error: delete() called on undefined index or empty List");
      --length;
      cursor.prev.next = cursor.next;
      cursor.next.prev = cursor.prev;
      cursor = null;
   }
   
   
   // Other Functions ---------------------------------------------------------

   // toString()
   // Overrides Object's toString method. Returns a String
   // representation of this List consisting of a space
   // separated sequence of integers, with front on left.
   public String toString(){
      String out = new String();
      Node N = front;
      
      for(N = front; N != null; N=N.next){
         out = out + String.valueOf(N.data) + " ";
      }
      return out;
   }

   // copy()
   // Returns a new List representing the same integer sequence as this
   // List. The cursor in the new list is undefined, regardless of the
   // state of the cursor in this List. This List is unchanged.
   List copy(){
      List L = new List();
      Node N = this.front;

      while( N!=null ){
         L.append(N.data);
         N = N.next;
      }
      return L;
   }
}